ambry-tools/src/main/java/com.github.ambry/store/DiskReformatter.java

```diff
@@ -0,0 +1,281 @@
+    for (ReplicaId replicaId : replicaIds) {
+      if (replicaId.getDiskId().getMountPath().equals(diskMountPath)) {
+        replicasOnDisk.add(replicaId);
+      }
+    }
+    if (replicasOnDisk.size() == 0) {
+      throw new IllegalArgumentException("There are no replicas on " + diskMountPath + " of " + dataNodeId);
+    }
+    replicasOnDisk.sort(Comparator.comparingLong(ReplicaId::getCapacityInBytes));
+    logger.info("Found {} on {}", replicasOnDisk, diskMountPath);
+
+    // move the last replica id (the largest one) to scratch space
+    ReplicaId toMove = replicasOnDisk.get(replicasOnDisk.size() - 1);
+    String partIdString = toMove.getPartitionId().toString();
+    File scratchSrc = new File(toMove.getReplicaPath());
+    File scratchTgt = new File(scratch, partIdString + RELOCATED_DIR_NAME_SUFFIX);
+    if (scratchTgt.exists()) {
+      throw new IllegalStateException(scratchTgt + " already exists");
+    }
+    logger.info("Moving {} to {}", scratchSrc, scratchTgt);
+    delete(scratchTgt);
+    FileUtils.moveDirectory(scratchSrc, scratchTgt);
+
+    // reformat each store, except the one moved, one by one
+    for (int i = 0; i < replicasOnDisk.size() - 1; i++) {
+      ReplicaId replicaId = replicasOnDisk.get(i);
+      partIdString = replicaId.getPartitionId().toString();
+      File src = new File(replicaId.getReplicaPath());
+      File tgt = new File(replicaId.getMountPath(), partIdString + UNDER_REFORMAT_DIR_NAME_SUFFIX);
+      logger.info("Copying {} to {}", src, tgt);
+      copy(partIdString, src, tgt, replicaId.getCapacityInBytes());
+      delete(src);
+      if (!tgt.renameTo(src)) {
+        throw new IllegalStateException("Could not rename " + tgt + " to " + src);
+      }
+      logger.info("Done reformatting {}", replicaId);
+    }
+
+    // reformat the moved store
+    copy(toMove.getPartitionId().toString(), scratchTgt, scratchSrc, toMove.getCapacityInBytes());
+    delete(scratchTgt);
+    logger.info("Done reformatting {}", toMove);
+  }
+
+  /**
+   * Copy the partition at {@code src} to {@code tgt} using a {@link StoreCopier}.
+   * @param storeId the name/id of the {@link Store}.
+   * @param src the location of the partition to be copied
+   * @param tgt the location where the partition has to be copied to
+   * @param capacityInBytes the capacity of the partition.
+   * @throws Exception
+   */
+  private void copy(String storeId, File src, File tgt, long capacityInBytes) throws Exception {
+    try (StoreCopier copier = new StoreCopier(storeId, src, tgt, capacityInBytes, fetchSizeInBytes, storeConfig,
+        new MetricRegistry(), storeKeyFactory, diskIOScheduler, Collections.EMPTY_LIST, time)) {
+      copier.copy(new StoreFindTokenFactory(storeKeyFactory).getNewFindToken());
+    }
+    // verify that the stores are equivalent
+    File[] replicas = {src, tgt};
+    if (!consistencyChecker.checkConsistency(replicas)) {
```

Comment:
if there are inconsistencies or any other errors, are you making sure that things are left in a good state as much as possible (that is there may be extraneous files for debugging and such, but we ensure that there is no data loss or availability issues - the server can be started successfully?)